x-pack/metricbeat/module/gcp/metrics/dataproc/metadata.go

```diff
@@ -0,0 +1,208 @@
+
+		if metadata.machineType != "" {
+			lastIndex := strings.LastIndex(metadata.machineType, "/")
+			_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudMachineTypeKey, metadata.machineType[lastIndex+1:])
+		}
+
+		metadata.Metrics = metadataCollectorData.Labels[gcp.LabelMetrics]
+		metadata.System = metadataCollectorData.Labels[gcp.LabelSystem]
+
+		if metadata.User != nil {
+			metadataCollectorData.Labels[gcp.LabelUser] = metadata.User
+		}
+	}
+
+	return metadataCollectorData, nil
+}
+
+// instanceMetadata returns the labels of an instance
+func (s *metadataCollector) instanceMetadata(ctx context.Context, instanceID, region string) (*dataprocMetadata, error) {
+	cluster, err := s.instance(ctx, instanceID)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to get data from instance '%s' %w", instanceID, err)
+	}
+
+	metadata := &dataprocMetadata{
+		clusterID: instanceID,
+		region:    region,
+	}
+
+	if cluster == nil {
+		s.logger.Debugf("couldn't get instance '%s' call ListInstances API", instanceID)
+		return metadata, nil
+	}
+
+	if cluster.ClusterName != "" {
+		parts := strings.Split(cluster.ClusterName, "/")
+		metadata.clusterName = parts[len(parts)-1]
+	}
+
+	if cluster.Labels != nil {
+		metadata.User = cluster.Labels
+	}
+
+	return metadata, nil
+}
+
+// instance returns data from an instance ID from the cache.
+func (s *metadataCollector) instance(ctx context.Context, instanceID string) (*dataproc.Cluster, error) {
+	s.getInstances(ctx)
+
+	instance, ok := s.clusters[instanceID]
+	if ok {
+		return instance, nil
+	}
+
+	s.clusters = make(map[string]*dataproc.Cluster)
+
+	return nil, nil
+}
+
+func (s *metadataCollector) instanceID(ts *monitoringpb.TimeSeries) string {
+	if ts.Resource != nil && ts.Resource.Labels != nil {
+		return ts.Resource.Labels["cluster_uuid"]
+	}
+
+	return ""
+}
+
+func (s *metadataCollector) instanceRegion(ts *monitoringpb.TimeSeries) string {
+	if ts.Resource != nil && ts.Resource.Labels != nil {
+		return ts.Resource.Labels["region"]
+	}
+
+	return ""
+}
+
+func (s *metadataCollector) getInstances(ctx context.Context) {
+	if len(s.clusters) > 0 {
+		return
+	}
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+
+	regionsToQuery := s.regions
+	if len(regionsToQuery) == 0 {
+		regionsToQuery = []string{"africa-south1", "asia-east1", "asia-east2", "asia-northeast1", "asia-northeast2", "asia-northeast3", "asia-south1", "asia-south2", "asia-southeast1", "asia-southeast2", "australia-southeast1", "australia-southeast2", "europe-central2", "europe-north1", "europe-north2", "europe-southwest1", "europe-west1", "europe-west10", "europe-west12", "europe-west2", "europe-west3", "europe-west4", "europe-west6", "europe-west8", "europe-west9", "me-central1", "me-central2", "me-west1", "northamerica-northeast1", "northamerica-northeast2", "northamerica-south1", "southamerica-east1", "southamerica-west1", "us-central1", "us-east1", "us-east4", "us-east5", "us-south1", "us-west1", "us-west2", "us-west3", "us-west4"}
```

Comment:
if new regions get added we end having to maintain this list. is there a way to pull 'all regions' from the API?

Reply:
Since dataproc relies heavily on compute, we could use [RegionsClient.List](https://pkg.go.dev/cloud.google.com/go/compute@v1.29.0/apiv1#RegionsClient.List). I it and it outputs the exact regions that are currently hard coded.

Reply:
i think that would be preferable to maintaining the list here.

Reply:
I'm also voting in favor of using the endpoint instead of maintaining this list.